File: 2013/Lunch & Learn/IntroToOOP/VB.NET/IntroToOOP/Example3/Animal.vb
Namespace IntroToOOP.Example3
    Public Class Animal

        Public Sub New()

        End Sub

        Public Sub New(name As String, color As String)
            Me.Name = name
            Me.Color = color
        End Sub

        Public Name As String
        Public Color As String

        Public Sub wagsTail()
            Console.WriteLine("{0} wags tail...", Name)
        End Sub

        Public Overridable Sub Speaks()
            Console.WriteLine("{0} grunts...", Name)
        End Sub
    End Class

    Public Class Cat
        Inherits Animal

        Public Sub New()
            Me.Name = "Sue"
            Me.Color = "Black"
        End Sub

        Public Sub New(name As String, color As String)
            MyBase.Name = name
            MyBase.Color = color
        End Sub

        Public Overrides Sub Speaks()
            MyBase.Speaks()
            Console.WriteLine("{0} purrs...", Me.Name)
        End Sub

    End Class

    Public Class OohCat
        Inherits Animal

        Public Overrides Sub Speaks()
            Console.WriteLine("{0} raises hand to mouth slowly and says OOOOOHHH!", Name)
        End Sub
    End Class

    Public Class Dog
        Inherits Animal

        Public Overrides Sub Speaks()
            Console.WriteLine("{0} barks and howls...", Name)
        End Sub

    End Class
End Namespace

File: 2013/Lunch & Learn/IntroToOOP/C#/Example3/Animal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace IntroToOOP.Example3
{
    public class Animal
    {
        public Animal()
        {
           
        }

        /* Consctructor with parameters (different signature) */
        public Animal(string name, string color)
        {
            Name = name;
            Color = color;
        }

        /* Public properties to allow the return and modification of the private members
         * Allow the State to be changed from outside the class*/
        public string Name { get; set;}

        public string Color { get; set;}           

        public void wagsTail()
        {
            Console.WriteLine("{0} wags tail...", Name);
        }

        public virtual void Speaks()
        {
            Console.WriteLine("{0} grunts...", Name);
        }

    }

    public class Cat : Animal
    {
        /* Constructor */
        public Cat() : base()
        {
            this.Name = "Sue";
            this.Color = "Black";
        }

        /* Consctructor with parameters (different signature) */
        public Cat(string name, string color) : base(name, color)
        {
            this.Name = name;
            this.Color = color;
        }
      
        /* Public properties to allow the return and modification of the private members
         * Allow the State to be changed from outside the class*/        
        public override void Speaks()
        {
            base.Speaks();
            Console.WriteLine("{0} purrs...", Name);
        }

    }

    public class OohCat : Animal
    {        
        public override void Speaks()
        {
            Console.WriteLine("{0} raises hand to mouth slowly and says OOOOOHHH!", Name);
        }
    }

    public class Dog : Animal
    {
        public override void Speaks()
        {
            Console.WriteLine("{0} barks and howls...", Name);
        }

    }


    
}
